### DISCUSSION TITLE
Custom type registry that complies with the PHPStan types

### DISCUSSION THREAD

**Discussion by NormySan** [webonyx/graphql-php#1371] [General]
Hey all! I've implemented a custom type registry that contains all of the registered types to be able to use dependency injection and to have these types only be created once. Similar to how it's done in the documentation.

This means that I have a get method on the registry to get a type from the registry like this:

```PHP
/**
 * Get a type instance based on class name or type name.
 *
 * @param string|class-string $classNameOrName the name of the class or type
 */
public function get(string $classNameOrName): ObjectType|InputObjectType|EnumType|ScalarType|UnionType
{
    if (isset($this->typesByName[$classNameOrName])) {
        return $this->typesByName[$classNameOrName];
    }

    if (isset($this->typesByClass[$classNameOrName])) {
        return $this->typesByClass[$classNameOrName];
    }

    throw new InvalidArgumentException("The {$classNameOrName} type has not been registered in the type registry.");
}
```

It's then used like this example where I create classes that provide all types and get the type registry injected to be able to use other types in the schema.

```PHP
class ServiceBookingFiltersInputType implements TypeInterface
{
    public function getType(TypeRegistry $registry): InputObjectType
    {
        return new InputObjectType([
            'name' => 'ServiceBookingFiltersInput',
            'fields' => fn () => [
                'customer' => [
                    'type' => Type::string(),
                    'defaultValue' => null,
                ],
                'service' => [
                    'type' => Type::id(),
                    'defaultValue' => null,
                ],
                'status' => [
                    'type' => $registry->get(ServiceBookingStatusType::class),
                    'defaultValue' => null,
                ],
            ],
        ]);
    }
}
```

This creates a problem now that I've added PHPStan to the project tough since the "InputObjectType" only expects the fields to contain types that are of "Type&InputType" but the get method can return any number of types since it can contain any kind of output and input types.

Has anyone had a similar problem and how did you solve it? I would prefer to keep only the "get" method instead of creating different methods for each kind of type.

**Comment by spawnia** [comment]
I can only see this working with generics, where your function accepts a `class-string<T>` and returns a type `T`.

**Reply by NormySan** [reply]
Thanks, didn't think of generics, might have to rebuild it tough since I wrap the types in a class that then generates the type so can't really use the type of that class as the return value.
